lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -891,6 +891,104 @@ OpFoldResult OrOp::fold(ArrayRef<Attribute> constants) {
       constants, [](APInt &a, const APInt &b) { a |= b; });
 }
 
+/// Merge concats in an or op when they have non-conflicting fields.
+static bool canonicalizeOrExclusiveConcat(OrOp op, size_t concatIdx1,
+                                          size_t concatIdx2,
+                                          PatternRewriter &rewriter) {
+  assert(concatIdx1 < concatIdx2 && "concatIdx1 must be < concatIdx2");
+
+  auto inputs = op.inputs();
+  auto concat1 = inputs[concatIdx1].getDefiningOp<ConcatOp>();
+  auto concat2 = inputs[concatIdx2].getDefiningOp<ConcatOp>();
+
+  assert(concat1 && concat2 && "expected indexes to point to ConcatOps");
+
+  auto knownBits1 = computeKnownBits(inputs[concatIdx1]);
+  auto knownBits2 = computeKnownBits(inputs[concatIdx2]);
+  auto unknownBits1 = ~(knownBits1.Zero | knownBits1.One);
+  auto unknownBits2 = ~(knownBits2.Zero | knownBits2.One);
+  bool anyUnknownBitsOverlap = !(unknownBits1 & unknownBits2).isNullValue();
+  bool anyConstantOne = !(knownBits1.One | knownBits2.One).isNullValue();
```

Comment:
I'd recommend taking a slightly different approach on this, which would make this a bit simpler and more general, building on the fact that `rewriter.createOrFold` autofolds and that canonicalize itself is iterative.

Instead of using computeKnownBits, just scan both concats to see if either operand has any constants as operands to the concat.  If so, turn:

`or (concat ..., concat ...)`

into: `concat(or ..., or ..., or ..., or ...)` where the chunks of the result are the maximum contiguous ranges from the two concats you have in.  When folding the nested or's, extracts from constants will automatically fold into smaller constants, and any "or" of 0 or 111 will fold automatically.

Does this sort of approach make sense to you?

Reply:
by max contiguous chunks, what I mean is that something like `concat(a_3bit, b_2bit)|concat(cst_2bit, d_3bit)` would turn into:

```
concat(or(extract(a_3bit, 2..1), cst_2bit),
              or(extract(a_3bit, 0...0), extract(d_3bit, 2..2)),
              or(b_2bit, extract(d_3bit, 1..0)))
```

Each of those things would nicely fold themselves.

Reply:
> where the chunks of the result are the maximum contiguous ranges from the two concats you have in.

Just to double check my understanding, are you saying we do this in the following example?

```
or(concat(x: i2, 0: i6), concat(0: i5, y: i3))
->
concat(or(x: i2, 0: i2), or(0: i3, 0: i3), or(0: i3, y: i3))
```

And in the case where x and y overlap, then we should extract pieces of the unknown values? For example:

```
or(concat(x: i6, 0: i2), concat(0: i1, y: i7))
->
// Using Verilog slice notation
concat(or(x[5:5], 0): i1, or(x[4:0], y[6:2]): i5, or(0, y[1:0]): i2)
```

If so, yeah, that makes sense to me, and I like that it allows the constant folding of the individual `or`s to happen automatically. I can work on making these changes.

Reply:
Somehow my browser tab only showed your first comment but not the second when I wrote up that comment. Sounds like I'm understanding it correctly, so let me work on that.

Reply:
Thanks, this worked out very well, and I've pushed up an amended commit that takes this approach.